primitives/fee-handler/src/fee_handler.rs

```diff
@@ -0,0 +1,370 @@
+
+	W: WithdrawCredit<
+		AccountId = AccountId,
+		Assets = Assets,
+		Credit = fungibles::Credit<AccountId, Assets>,
+	>,
+
+	Affiliate: DistributeFee<AccountId = W::AccountId, Balance = W::Balance>,
+	Tournament:
+		DistributeFee<AccountId = AccountId, Balance = W::Balance, MaxDistributions = ConstU32<1>>,
+{
+	type AccountId = AccountId;
+	type AssetId = W::AssetId;
+	type Balance = W::Balance;
+	type AffiliateFeeIdentifier = Affiliate::FeeIdentifier;
+	type TournamentFeeIdentifier = Tournament::FeeIdentifier;
+
+	fn withdraw_and_pay_fees(
+		payer: &Self::AccountId,
+		payment_asset: Self::AssetId,
+		base_fee: Self::Balance,
+		tournament_id: &Self::TournamentFeeIdentifier,
+		affiliate_id: &Self::AffiliateFeeIdentifier,
+		treasury_pot: &Self::AccountId,
+	) -> Result<(), DispatchError> {
+		// The credit may be in any asset as implemented by `WithdrawAsset`.
+		let fee_credit = W::withdraw_credit(payer, payment_asset, base_fee)?;
+
+		let remaining_credit =
+			Self::try_propagate_tournament_fee(fee_credit, payer, tournament_id)?;
+
+		let remaining_credit2 =
+			Self::try_propagate_chain_fee(remaining_credit, payer, affiliate_id)?;
+
+		Self::deposit_into_treasury(treasury_pot, remaining_credit2)
+	}
+
+	fn withdraw_and_deposit_into_treasury(
+		who: &Self::AccountId,
+		asset_id: Self::AssetId,
+		treasury_pot: &Self::AccountId,
+		amount: Self::Balance,
+	) -> Result<(), DispatchError> {
+		let credit = W::withdraw_credit(who, asset_id, amount)?;
+		Self::deposit_into_treasury(treasury_pot, credit)
+	}
+}
+
+impl<AccountId, Assets, W, Affiliate, Tournament>
+	AssetGameFeeHandler<AccountId, Assets, W, Affiliate, Tournament>
+where
+	Assets: fungibles::Inspect<AccountId, Balance = W::Balance, AssetId = W::AssetId>
+		+ fungibles::Balanced<AccountId>,
+
+	W: WithdrawCredit<
+		AccountId = AccountId,
+		Assets = Assets,
+		Credit = fungibles::Credit<AccountId, Assets>,
+	>,
+
+	Affiliate: DistributeFee<AccountId = W::AccountId, Balance = W::Balance>,
+	Tournament:
+		DistributeFee<AccountId = AccountId, Balance = W::Balance, MaxDistributions = ConstU32<1>>,
+{
+	/// Distributes an already withdrawn `fee_credit` to the affiliates of `account`.
+	///
+	/// Returns the remaining `fee_credit` after this operation.
+	fn try_propagate_chain_fee(
+		fee_credit: W::Credit,
+		account: &W::AccountId,
+		identifier: &Affiliate::FeeIdentifier,
+	) -> Result<W::Credit, DispatchError> {
+		let mut final_fee = fee_credit;
+
+		if let Some(a) = Affiliate::distribute_fee(final_fee.peek(), account, identifier) {
+			for allocation in a {
+				if allocation.amount > 0_u32.into() {
+					let affiliate_fee = final_fee.extract(allocation.amount);
+
+					if let Err(credit) = W::Assets::resolve(&allocation.beneficiary, affiliate_fee)
+					{
+						log::error!(
+							"Could not deposit to affiliate account, it probably doesn't exist."
+						);
+
+						// Reabsorb the credit; it can still be used.
+						let _ = final_fee.subsume(credit).defensive();
+					}
```

Comment:
Shouldn't we just let the method fail if any of the allocations cannot be distributed? They way it is implemented right now, it would be quite difficult to trace any error that happens in `resolve`